### PR TITLE
fix: panic index out of range for invalid series keys

### PR DESCRIPTION
**Pull request by jdockerty**
Panics can be avoided by checking the validity of the return from `ReadSeriesKeyLen`, at the moment the size check resulting from this function is entirely discarded. The contained call to [`binary.Uvarint`](https://pkg.go.dev/encoding/binary#Uvarint) contains an implicit error, depending on the returned value:

> If an error occurred, the value is 0 and the number of bytes n is <= 0

In every occurrence of `ReadSeriesKeyMeasurement`, the `ReadSeriesKeyLen` function is called prior to it, determined by glancing over the output of `rg "ReadSeriesKeyMeasurement" -B 5` across the full codebase. 

This PR refactors the flow slightly to include `ReadSeriesKeyLen` returned value checks and the appropriate action to avoid panicking when trying to access memory out of bounds, it has a knock-on effect such that the `ParseSeriesKey` function can return `nil` at another point which indicates an invalid result and the appropriate action should be taken.

Closes: https://github.com/influxdata/influxdb/issues/24454, https://github.com/influxdata/influxdb/issues/24469, https://github.com/influxdata/influxdb/issues/24432, https://github.com/influxdata/influxdb/issues/17409, https://github.com/influxdata/influxdb/issues/20245, https://github.com/influxdata/influxdb/issues/23266

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: tsdb/series_file.go [hunk 1]
```diff
@@ -445,8 +448,11 @@ func CompareSeriesKeys(a, b []byte) int {
 	}
 
 	// Read total size.
-	_, a = ReadSeriesKeyLen(a)
-	_, b = ReadSeriesKeyLen(b)
+	szA, a := ReadSeriesKeyLen(a)
+	szB, b := ReadSeriesKeyLen(b)
+	if szA == 0 && szB == 0 {
+		return 0
+	}
```
Comment (davidby-influx) [inline tsdb/series_file.go:455]: Could this be a more complete test?  Something like this?  (Caveat, this may not be correct logic, 
 but I want to avoid the calls to `ReadSeriesKeyMeasurement` below for a zero length key)
```
    switch cmp := szA - szB; cmp {
    case szA != 0 && szB != 0:
        break
    case cmp > 0:
        return 1
    case cmp < 0:
        return -1
    case cmp == 0:
        return 0
    default:
    }
    // Read names.
    name0, a := ReadSeriesKeyMeasurement(a)
    name1, b := ReadSeriesKeyMeasurement(b)
```

File: tsdb/series_file.go [hunk 2]
```diff
@@ -413,7 +413,10 @@ func ParseSeriesKeyInto(data []byte, dstTags models.Tags) ([]byte, models.Tags)
 // different length and capacity to those provided.
 func parseSeriesKey(data []byte, dst models.Tags) ([]byte, models.Tags) {
 	var name []byte
-	_, data = ReadSeriesKeyLen(data)
+	sz, data := ReadSeriesKeyLen(data)
+	if sz == 0 && len(data) == 0 {
```
Comment (davidby-influx) [inline tsdb/series_file.go:417]: I think this test should simply be `if len(data) == 0 {`

File: tsdb/index/tsi1/log_file.go [hunk 3]
```diff
@@ -712,7 +712,10 @@ func (f *LogFile) execSeriesEntry(e *LogEntry) {
 	deleted := e.Flag == LogEntrySeriesTombstoneFlag
 
 	// Read key size.
-	_, remainder := tsdb.ReadSeriesKeyLen(seriesKey)
+	sz, remainder := tsdb.ReadSeriesKeyLen(seriesKey)
+	if sz == 0 && len(remainder) == 0 {
```
Comment (davidby-influx) [inline tsdb/index/tsi1/log_file.go:716]: Should this be `if len(remainder) == 0 {`

File: tsdb/index.go [hunk 4]
```diff
@@ -2918,12 +2933,18 @@ func (is IndexSet) tagValuesByKeyAndExpr(auth query.FineAuthorizer, name []byte,
...
+				continue
+			}
 			if !auth.AuthorizeSeriesRead(database, name, tags) {
 				continue
 			}
 		}
 
-		_, buf = ReadSeriesKeyLen(buf)
+		sz, buf := ReadSeriesKeyLen(buf)
+		if sz == 0 && len(buf) == 0 {
```
Comment (davidby-influx) [inline tsdb/index.go:2945]: should this just be `if len(buf) == 0 {` ?

File: tsdb/index.go [hunk 5]
```diff
@@ -2442,6 +2454,9 @@ func (is IndexSet) MeasurementSeriesKeysByExpr(name []byte, expr influxql.Expr)
 		}
 
 		name, tags := ParseSeriesKey(seriesKey)
+		if name == nil && tags == nil {
```
Comment (davidby-influx) [inline tsdb/index.go:2457]: Maybe an `||` instead of `&` ?

File: tsdb/series_partition.go [hunk 6]
```diff
@@ -263,7 +263,10 @@ func (p *SeriesPartition) CreateSeriesListIfNotExists(keys [][]byte, keyPartitio
 		newIDs[string(key)] = id
 		newKeyRanges = append(newKeyRanges, keyRange{id, offset})
 		if tracker.AddedMeasurementSeries != nil {
-			_, remainder := ReadSeriesKeyLen(key)
+			sz, remainder := ReadSeriesKeyLen(key)
+			if sz == 0 && len(remainder) == 0 {
+				continue
+			}
 			measurement, _ := ReadSeriesKeyMeasurement(remainder)
```
Comment (davidby-influx) [inline tsdb/series_partition.go:270]: Do we need to check if `measurement` has zero length here?

File: tsdb/index.go [hunk 7]
```diff
@@ -2918,12 +2933,18 @@ func (is IndexSet) tagValuesByKeyAndExpr(auth query.FineAuthorizer, name []byte,
 
 		if auth != nil {
 			name, tags := ParseSeriesKey(buf)
+			if name == nil && tags == nil {
```
Comment (davidby-influx) [inline tsdb/index.go:2936]: Would it be enough to bail if `name == nil`?

**Review by davidby-influx — COMMENTED** [review]
The overall comment I have is that the tests when `ReadSeriesKeyLen` return a zero size or an empty buffer should be stricter in preventing the next operation, but you need to be sure they are correct.  For instance, can we have a key with no tags?  What does that produce out of the various parsing functions? I believe it is possible, so depending on what happens after the parsing depends on what's being done; are we processing the tags, or only the key name?  

So, take a look at each test you've added and see what comes next, then make the test as simple and robust as possible.

File: tsdb/series_file.go (hunk 1 above)
Comment (jdockerty) [reply tsdb/series_file.go:455]: This makes sense 👍 

I was wondering too shouldn't this perhaps be consolidated into the check above?

https://github.com/influxdata/influxdb/blob/2556e3875f2f836e15f96c579292a58015afd8eb/tsdb/series_file.go#L440-L448

I'm failing the see the difference between the initial `if len(a) == 0 ...` portion which allegedly handles `nil` keys in a similar manner to what we're trying here, but then the call into `ReadSeriesKeyLen` below which would provide a `szA = 0` _should_ have already been skipped over from the prior check and early return.

File: tsdb/series_file.go (hunk 2 above)
Comment (jdockerty) [reply tsdb/series_file.go:417]: I think you're right, I believe I've been a bit heavy-handed in also enforcing the `sz && len(data)`, as if `len(data) == 0`, then the returned size was also 0 anyway.

File: tsdb/index.go (hunk 5 above)
Comment (jdockerty) [reply tsdb/index.go:2457]: I think this might be the case of having `if name == nil` alone, right?

As if the series key was `nil` then that is invalid, _but_ you can have a series key with simply no tags (`nil`) and still have a valid measurement.

I'll hold off on altering this one for now, just in case I'm not understanding it properly and it'll avoid the change back/forth.

File: tsdb/index.go [hunk 8]
```diff
@@ -2454,7 +2454,7 @@ func (is IndexSet) MeasurementSeriesKeysByExpr(name []byte, expr influxql.Expr)
 		}
 
 		name, tags := ParseSeriesKey(seriesKey)
-		if name == nil && tags == nil {
+		if name == nil || tags == nil {
```
Comment (davidby-influx) [inline tsdb/index.go:2457]: Are you sure we don't ever want to add a key with no tags here?  I'm not sure, I just wanted you to verify exactly what was the right test. exactly what

**Review by davidby-influx — COMMENTED** [review]
Onbly one question remaining on the conditionals and early abandonment.

File: tsdb/index.go (hunk 7 above)
Comment (jdockerty) [reply tsdb/index.go:2936]: Agreed!

Similar to what we discussed yesterday, there is a similar prior check above this part so it _should_ be caught already. A `name == nil` on the parsing is a definite invalid value we want to skip.

Just above the highlighted code snippet for this comment :point_down: 
```go
		buf := is.SeriesFile.SeriesKey(e.SeriesID)
		if len(buf) == 0 {
			continue
		}
```

File: tsdb/series_file_test.go [hunk 9]
```diff
@@ -50,20 +50,33 @@ func TestParseSeriesKeyInto(t *testing.T) {
...
+			name:        "valid series key with tags",
+			seriesKey:   tsdb.AppendSeriesKey(nil, []byte("foo"), models.NewTags(map[string]string{"tag1": "foo"})),
+			expectedKey: []byte("foo"),
+		},
+		{
+			name:        "valid series key with no tags",
+			seriesKey:   tsdb.AppendSeriesKey(nil, []byte("foo"), nil),
+			expectedKey: []byte("foo"),
+		},
 	}
```
Comment (davidby-influx) [inline tsdb/series_file_test.go:74]: I'd like two more tests here, with a zero-length `models.NewTags` created
- With an empty map
- With a `nil` map

We've had lots of bugs caused by different handling between `nil` and empty collections.

File: tsdb/index.go [hunk 10]
```diff
@@ -2454,7 +2454,9 @@ func (is IndexSet) MeasurementSeriesKeysByExpr(name []byte, expr influxql.Expr)
 		}
 
 		name, tags := ParseSeriesKey(seriesKey)
-		if name == nil || tags == nil {
+		// An invalid series key of 0 length should have been caught by the
+		// above check, but for extra safety we can pass over it here too.
+		if name == nil {
```
Comment (davidby-influx) [inline tsdb/index.go:2459]: I like extra safety.

**Review by davidby-influx — CHANGES_REQUESTED** [review]
A few more tests requested.

File: tsdb/series_file_test.go (hunk 9 above)
Comment (jdockerty) [reply tsdb/series_file_test.go:89]: Makes sense, I've expanded these now 👍

File: tsdb/series_file.go (hunk 1 above)
Comment (davidby-influx) [inline tsdb/series_file.go:455]: I believe you still need to fix this test up, to avoid calls to `ReadSeriesKeyMeasurement` with empty buffers.  Either the way I suggested or the way you suggested from the start of the function, or some combination, but I would like to see the operation halted if either key is zero length.

**Review by davidby-influx — CHANGES_REQUESTED** [review]
One unresolved issue we already discussed.

File: tsdb/series_file.go (hunk 1 above)
Comment (jdockerty) [reply tsdb/series_file.go:459]: I believe I've overcomplicated this when I was first reading it, those initial checks are okay and we can conduct something very similar because the result for our `sz` is only ever `0` on an error, which is bubbled up from [`binary.Uvarint`](https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/encoding/binary/varint.go;l=69-88):
```go
func Uvarint(buf []byte) (uint64, int) {
	var x uint64
	var s uint
	for i, b := range buf {
		if i == MaxVarintLen64 {
			// Catch byte reads past MaxVarintLen64.
			// See issue https://golang.org/issues/41185
			return 0, -(i + 1) // overflow
		}
		if b < 0x80 {
			if i == MaxVarintLen64-1 && b > 1 {
				return 0, -(i + 1) // overflow
			}
			return x | uint64(b)<<s, i + 1
		}
		x |= uint64(b&0x7f) << s
		s += 7
	}
	return 0, 0
}
```

I've pulled similar checks further down to the `ReadSeriesKeyLen` as well to cover it, it looks like similar logic was done on a tags check further down here too.

**Review by davidby-influx — APPROVED** [review]
LGTM - pass on to @gwossum for a second clean review and prepare for all the ports (1.11, main-2.x, 2.7)

**Review by davidby-influx — DISMISSED** [review]
LGTM!  Thanks for contributing.  Please get a review from @gwossum and if he approves, create issues for ports to main-2.x, 2.7, and 1.11 branches, then cherry-pick into those and put up the PRs.

File: cmd/influx_inspect/dumptsi/dumptsi.go [hunk 11]
```diff
@@ -244,6 +244,9 @@ func (cmd *Command) printSeries(sfile *tsdb.SeriesFile) error {
 			break
 		}
 		name, tags := tsdb.ParseSeriesKey(sfile.SeriesKey(e.SeriesID))
+		if name == nil {
+			continue
+		}
```
Comment (gwossum) [inline cmd/influx_inspect/dumptsi/dumptsi.go:249]: @jdockerty @davidby-influx Should we be checking for `len(name) == 0` instead of `name == nil`? Checking the `len` would also catch the case where the name is empty in addition to nil. Or is an empty name valid?

**Review by gwossum — COMMENTED** [review]
Assuming `name == nil` is the check we need to perform, I don't see any issues.

File: cmd/influx_inspect/dumptsi/dumptsi.go (hunk 11 above)
Comment (jdockerty) [reply cmd/influx_inspect/dumptsi/dumptsi.go]: For any reader's clarity, we discussed this offline and concluded that `len(name) == 0` was a safer/wider check to handle potential edge cases, whilst also still covering my initial intention.

I've altered this now 💯

**Review by davidby-influx — APPROVED** [review]
LGTM

**Review by gwossum — APPROVED** [review]
LGTM
